src/plugins/ui_actions/public/actions/action_internal.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Action, ActionContext, AnyActionDefinition } from './action';
+import { Presentable } from '../util/presentable';
+import { createActionStateContainer, ActionState } from './action_state_container';
+import { uiToReactComponent } from '../../../kibana_react/public';
+import { ActionContract } from './action_contract';
+import { ActionType } from '../types';
+
+export class ActionInternal<A extends AnyActionDefinition>
+  implements Action<ActionContext<A>>, Presentable<ActionContext<A>> {
+  constructor(public readonly definition: A) {}
+
+  public readonly id: string = this.definition.id;
+  public readonly type: ActionType = this.definition.type || '';
+  public readonly order: number = this.definition.order || 0;
+  public readonly MenuItem? = this.definition.MenuItem;
+  public readonly ReactMenuItem? = this.MenuItem ? uiToReactComponent(this.MenuItem) : undefined;
+  public readonly CollectConfig? = this.definition.CollectConfig;
```

Comment:
After working on UI components it seems to me that it would be better to have this collect config thingy on `ActionFactory` instead of `Action`. 

Because when we open a dialog for creating / editing a drilldown - user can **change** the type of action he is creating.

It means, that we either should be able to change action's type on a fly (and all other action definition things which depend on it) or just don't create any action when editing. 

So, I think, the way to go is to make that editor component to not have `Action` as Input and Output, but instead have something like as input and output all needed information to create an action. In our case: 
`
actionFactory
actionConfig
name
`
where actionFactory will have
``` 
actionFactory.createConfig() // default config
actionFactory.CollectConfig() // specific component for collecting it
actionFactory.isValid(name, config) // is info for creating an action is valid? 
```

So far I was following this direction in my components prs. 
#59032 #59139